docs: add probationary step to Experimental GEP phase (#2097)
kubernetes-sigs/gateway-api: geps/overview.md

stage.

### 4. Implement the GEP as "Experimental"

With the GEP marked as "Implementable", it is time to actually make those
proposed changes in our API. In some cases, these changes will be documentation
only, but in most cases, some API changes will also be required. It is important
that every new feature of the API is marked as "Experimental" when it is
introduced. Within the API, we use `<gateway:experimental>` tags to denote
experimental fields. In order to transition a GEP into `Experimental` **the
graduation criteria for the GEP MUST be filled out**, indicating all of the
currently known issues and dilemmas that need to be resolved in order to
graduate to `Standard`.

Before these changes are released, they will also need to be documented.
GEPs that have not been both implemented and documented before a release
cut off will be excluded from the release.

> ### 5. Probationary Period

Reviewer: It may help to provide a sentence or paragraph that describes the rationale for this - essentially trying to avoid the state of features getting stuck in experimental forever.

Author: Added in 2cc4acc93314fe0590d3afe82191ce571f023c6d

Any GEP in the `Experimental` phase is automatically under a "probationary
period" where the **maintainers will assess the ability and likelihood that

> the graduation criteria can be resolved within a given time period**. By

Reviewer: Tiny nit that applies here and elsewhere in the PR, I think this may be overusing bold formatting. The current doc does not have any bold text, and we've historically used it very sparingly throughout Gateway API docs.

Author: Changed in 2cc4acc93314fe0590d3afe82191ce571f023c6d

default that time period is **6 months**. Maintainers may select an alternative

> time duration for a probationary period if appropriate, and will document their

Reviewer: Maybe we should just add the following requirements for GEPs graduating to experimental:

1. Clearly document graduation criteria
2. Propose date after which the GEP will automatically be rejected if graduation criteria is not met

That leaves more of the power in the hands of the person pushing the GEP forward and more closely links the graduation criteria to a timeline.

Author: Changed in 2cc4acc93314fe0590d3afe82191ce571f023c6d

stage.

### 4. Implement the GEP as "Experimental"

With the GEP marked as "Implementable", it is time to actually make those
proposed changes in our API. In some cases, these changes will be documentation
only, but in most cases, some API changes will also be required. It is important
that every new feature of the API is marked as "Experimental" when it is
introduced. Within the API, we use `<gateway:experimental>` tags to denote
experimental fields.

Some other requirements must be met before marking a GEP `Experimental`:

- the graduation criteria to reach `Standard` MUST be filled out
- a proposed probationary period (see next section) must be included in the GEP
  and approved by maintainers.

Before changes are released they MUST be documented. GEPs that have not been
both implemented and documented before a release cut off will be excluded from
the release.

> Any GEP in the `Experimental` phase is automatically under a "probationary

Reviewer 2: Nit: maybe add

```
#### Probationary Period
```

above this line to make this easier to find when skimming?

Author: ```suggestion
#### Probationary Period

Any GEP in the `Experimental` phase is automatically under a "probationary
```

period" where it will come up for re-assessment if its graduation criteria are

> not met within a given time period. GEPs that wish to move into `Experimental`

Reviewer 2: Nit: should we suggest a default time period? or possibly a maximum beyond which you should expect pushback?

Author: I originally started with a default time period of 6 months but @robscott I think wanted it to be something that was proposed. I think there's a middle ground here where we suggest a time:

```suggestion
not met within a given time period (6 months is the suggested default). GEPs that wish to move into `Experimental`
```

stage.

### 4. Implement the GEP as "Experimental"

> 

Reviewer: Are these additional lines intentional?

With the GEP marked as "Implementable", it is time to actually make those
proposed changes in our API. In some cases, these changes will be documentation
only, but in most cases, some API changes will also be required. It is important
that every new feature of the API is marked as "Experimental" when it is
introduced. Within the API, we use `<gateway:experimental>` tags to denote
experimental fields.

> Some other requirements must be met before marking a GEP `Experimental`:
> - the graduation criteria to reach `Standard` MUST be filled out

Reviewer: new line needed for list formatting: https://deploy-preview-2097--kubernetes-sigs-gateway-api.netlify.app/geps/overview/

```suggestion
Some other requirements must be met before marking a GEP `Experimental`:

- the graduation criteria to reach `Standard` MUST be filled out
```

- a proposed probationary period (see next section) must be included in the GEP
  and approved by maintainers.

Before changes are released they MUST be documented. GEPs that have not been
both implemented and documented before a release cut off will be excluded from
the release.

> ### 5. Probationary Period

Reviewer: I'm not sure where this comment belongs, but I'm a bit uncomfortable with the "Probationary" term here and adding another phase. I think what we may actually want here is just to clarify that:

* Experimental is a temporary phase
* GEPs will have regular check ins and must show progress towards standard channel
* As part of graduating to experimental, deadlines will be defined for key milestones on the path to GA, including final graduation
* If any of those deadlines are missed, we may choose to either:
  * Remove the GEP from the experimental channel (and therefore the API)
  * Provide a one-time extension of certain deadlines as agreed by the community

Reviewer: I think I'm basically saying the same thing as you are, but the bits that were making me uncomfortable about the current wording were:

1. Is Probationary a phase in the GEP process? 
2. If so, do we now have two phases (Probationary and Experimental) that both land in experimental channel?
3. Does having a probationary state actually result in the "Experimental" state afterwards being viewed as more stable? IE once a GEP leaves probationary does it exist in Experimental channel without any risk of breaking changes or removal?

My suggestion above is just one possible way to address those questions, always open to other ideas.

Author: ```
I think what we may actually want here is just to clarify that:

    Experimental is a temporary phase
    GEPs will have regular check ins and must show progress towards standard channel
    As part of graduating to experimental, deadlines will be defined for key milestones on the path to GA, including final graduation
    If any of those deadlines are missed, we may choose to either:
        Remove the GEP from the experimental channel (and therefore the API)
        Provide a one-time extension of certain deadlines as agreed by the community
```

I'm in a difficult spot in terms of how to progress because it occurs to me this is _exactly_ what's being covered with the current language?

> Is Probationary a phase in the GEP process?
> If so, do we now have two phases (Probationary and Experimental) that both land in experimental channel?
> Does having a probationary state actually result in the "Experimental" state afterwards being viewed as more stable?

"Probationary" as proposed here is part of the "GEP process" not the GEP phases (e.g. implementable, experimental, e.t.c.), in the same way that "Document Implementation Details" is not a phase, just to clarify that. The steps here do not all align with specific phases. Probation is being proposed as an _aspect_ of the `Experimental` phase.

---

In regards to your discomfort with the "probation" language I have the opposite feeling: I am strongly in favor of the language as I think it is accurate and cautionary and I like it front and center. Do you think you can be persuaded?

Author: After talking about it with Rob in slack, I think we have a compromise. Latest changes have merged the two "steps" together.

    practical tests and experiments which are intended to help us learn and
    iterate on the GEP. These can include distributing content, but not under
    any release channel.

> * Q: Should I implement support for `Experimental` channel features?

Author: Could you please point me to the exact file and line where you would like to see this at?

> * Q: Should I implement support for `Experimental` channel features?

Reviewer: Didn't have a specific line in mind, also not intended to be blocking, fine to leave for a follow up.

  * A: Ultimately one of the main ways to get something into `Standard` is for
    it to mature through the `Experimental` phase, so we really _need_ people
    to implement these features and provide feedback in order to have progress.

>     That said, the graduation of a feature past `Experimental` is not a forgone
>     conclusion: **you MUST not implement an experimental feature unless your
>     implementation has a good way to indicate the potential transcience of that
>     feature to its users. You MUST also have a "spin down" plan in place in the
>     event that the GEP never moves forward to Standard and instead ends up being
>     Rejected** (e.g. place the feature behind a feature-gate system).

Reviewer: Although I agree with what you're saying here, it might be helpful to tweak the language to be a bit less strong:
```suggestion
    That said, the graduation of a feature past `Experimental` is not a forgone
    conclusion. Before implementing an experimental feature, you should:

    * Clearly document that support for the feature is experimental and may disappear in the future
    * Have a plan in place for how you would handle the removal of this feature from the API.
```
